CASSANDRA-19962: CEP-44 Kafka integration for Cassandra CDC using Sidecar

jberragan opened the pull request:
This is the initial commit for CEP-44 to introduce a standalone CDC module into the Analytics project. This module provides the foundation for CDC in the Apache Cassandra Sidecar.

This module provides:
- a standalone Cdc class as the entrypoint for initializing CDC.
- pluggable interfaces for: listing and reading commit log segments for a token range, persisting and reading CDC state, providing the Cassandra table schema, optionally reading values from Cassandra.
- read and deserialize commit log mutations.
- reconcile and de-duplicate mutations across replicas.
- serialize CDC state into a binary object for persistence.
- a layer for converting Cassandra mutations into a standard consumable format.

jberragan committed:
CEP-44 Kafka integration for Cassandra CDC using Sidecar

This is the initial commit for CEP-44 to introduce a standalone CDC module into the Analytics project. This module provides the foundation for CDC in the Apache Cassandra Sidecar.

This module provides:
- a standalone Cdc class as the entrypoint for initializing CDC.
- pluggable interfaces for: listing and reading commit log segments for a token range, persisting and reading CDC state, providing the Cassandra table schema, optionally reading values from Cassandra.
- read and deserialize commit log mutations.
- reconcile and de-duplicate mutations across replicas.
- serialize CDC state into a binary object for persistence.
- a layer for converting Cassandra mutations into a standard consumable format.

Co-authored-by: James Berragan <jberragan@apple.com>
Co-authored-by: Bernardo Botella Corbi <bernardo.botella@apple.com>
Co-authored-by: Yifan Cai <ycai@apache.org>
Co-authored-by: Jyothsna Konisa <jkonisa@apple.com>
Co-authored-by: Dinesh Joshi <djoshi@apache.org>

jberragan committed:
Make CDC version independent
Simplify message conversion generics

jberragan committed:
Fix tests

jberragan committed:
Relax epoch condition to reduce race chance

jberragan commented:
Green CI: https://app.circleci.com/pipelines/github/jberragan/cassandra-analytics/136/workflows/8820f428-ad79-42d8-8b44-eaceb471297c

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/cdc/state/CdcState.java:
```diff
@@ -0,0 +1,423 @@
+/**
+ * The CdcState object describes the cdc state that must be persisted between microbatches to make cdc recoverable.
+ * The key elements are of the state are:
+ * 1) the commit log markers, indiciating the maximal position reached for a given CassandraInstance.
```
```suggestion
 * The CdcState object describes the cdc state that must be persisted between micro-batches to make cdc recoverable.
 * The key elements are of the state are:
 * 1) the commit log markers, indicating the maximal position reached for a given CassandraInstance.
```

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/cdc/api/CommitLogMarkers.java:
```diff
@@ -0,0 +1,400 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public interface CommitLogMarkers extends Serializable
```
Can you extract the classes in this file into multiple files? It is hard to read. 
Can you also avoid putting implementations in the interface? Maybe `abstract class` is more suitable here.

yifan-c commented on cassandra-analytics-common/src/main/java/org/apache/cassandra/cdc/msg/CdcEventBuilder.java:
```diff
@@ -0,0 +1,212 @@
+    public CdcEvent build()
+    {
+        validateRangeTombstoneMarkers();
+        return new CdcEvent(
+        partitionKeys,
+        clusteringKeys,
+        staticColumns,
+        valueColumns,
+        maxTimestampMicros,
+        timeToLive,
+        tombstonedCellsInComplex,
+        rangeTombstoneList,
+        track,
+        trackingId,
+        keyspace,
+        table,
+        kind,
+        cassandraSource
+        );
```
nit: generally easier to read with passing the builder instance to the constructor, instead of having a long parameter list.

yifan-c reviewed (commented):
Looks good to me in general. Some nit comments.

jberragan committed:
Review comments:
- Extract CommitLogMarkers into standalone classes
- Use CdcEventBuilder in CdcEvent constructor
- Comment typo corrections

jberragan committed:
Other final cleanup

yifan-c reviewed (approved):
+1
No more comments from me.

dineshjoshi commented:
+1, thanks.
